rusoto/core/src/future.rs

```diff
@@ -141,140 +142,64 @@ lazy_static! {
 /// #   fn upload_part_copy(&self, _: UploadPartCopyRequest) -> RusotoFuture<UploadPartCopyOutput, UploadPartCopyError> { unimplemented!() }
 /// }
 /// ```
+pub(crate) type RusotoHandlerFuture<T, E> = Pin<Box<dyn Future<Output = Result<T, RusotoError<E>>> + Send>>;
+
+#[pin_project]
 pub struct RusotoFuture<T, E> {
-    state: Option<RusotoFutureState<T, E>>,
+    #[pin]
+    inner: RusotoHandlerFuture<T, E>,
+}
+
+async fn dispatch_rusoto_future<T, E>(
+    future: SignAndDispatchFuture,
+    handler: fn(HttpResponse) -> RusotoHandlerFuture<T, E>
+) -> Result<T, RusotoError<E>> {
+    let resp = future.await.map_err(RusotoError::SignAndDispatch)?;
+    handler(resp).await
 }
 
 pub fn new<T, E>(
-    future: Box<dyn TimeoutFuture<Item = HttpResponse, Error = SignAndDispatchError> + Send>,
-    handler: fn(HttpResponse) -> Box<dyn Future<Item = T, Error = RusotoError<E>> + Send>,
-) -> RusotoFuture<T, E> {
+    future: SignAndDispatchFuture,
+    handler: fn(HttpResponse) -> RusotoHandlerFuture<T, E>,
+) -> RusotoFuture<T, E>
+    where
+        T: 'static,
+        E: 'static
+{
     RusotoFuture {
-        state: Some(RusotoFutureState::SignAndDispatch { future, handler }),
+        inner: dispatch_rusoto_future(future, handler).boxed(),
     }
 }
 
 impl<T, E> RusotoFuture<T, E> {
```

Comment:
A thing that might have far-reaching impact, but ultimately worthwhile impact—I think you can remove the usage of `RusotoFuture` wholesale and instead have `async fn foo() -> Result<T, E>` or a `Box<dyn Future<...>>`.